### PR TITLE
rules: no-noninteractive-tabindex, no-redundant-roles, no-noninteractive-element-to-interactive-role

### PR DESCRIPTION
**Pull request by ematipico**
Implementation of the rules:
-  [no-noninteractive-tabindex.md](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-tabindex.md)
- [no-redundant-roles](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-redundant-roles.md)
- [no-noninteractive-element-to-interactive-role](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-element-to-interactive-role.md)

Refers to #341 

This is going to be a long one unfortunately! I guess we'd need to map more elements and roles and add more tests. For now I added enough tests to cover the combination of interactive roles, interactive elements, negative tabIndex and zero tabIndex. 

I'd like some feedback in the meantime!

### CODE REVIEW

**Comment by sebmck** [comment]
I have no idea how aria roles work so I'll defer review to someone else 😛

**Comment by ematipico** [comment]
@sebmck it's all documented here: https://www.w3.org/WAI/PF/aria/roles#article

Every role has a table with all the information we need. At the moment I'm covering the most used roles and HTML elements, that's why there only some.

But once have tests that cover the edge cases, the `roles` object can be filled with another PR with more cases

**Comment by ematipico** [comment]
The PR should be ready to be reviewed!

**Comment by sebmck** [comment]
cc @bitpshr @VictorHom

File: packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveElementToInteractiveRole.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	roles,
+} from "@romejs/compiler/lint/rules/ariaHelpers";
+
+export default {
+	name: "jsxA11YNoNoninteractiveElementToInteractiveRole",
+	enter(path: Path): TransformExitResult {
+		const {node} = path;
+
+		if (isJSXElement(node)) {
+			if (hasJSXAttribute(node, "role")) {
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveElementToInteractiveRole.ts:21]: Nit: These two `if` statements can be combined.

File: packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveElementToInteractiveRole.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+				if (
+					roleAttribute &&
+					roleAttribute.value &&
+					roleAttribute.value.type === "JSStringLiteral"
+				) {
+					const role = roles.get(roleAttribute.value.value);
+					if (role) {
+						if (!isElementInteractive(name) && isRoleInteractive(role)) {
+							path.context.addNodeDiagnostic(
+								node,
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveElementToInteractiveRole.ts:33]:
```suggestion
								roleAttribute,
```

This will increase the specificity of the node reported in the diagnostic.

File: packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+				 */
+				if (!hasJSXAttribute(node, "role")) {
+					const elementToRole = elementsToConcepts.get(elementName);
+					// the element is not part of any role, so it's an error
+					// e.g. div, span, etc.
+					if (elementToRole) {
+						// not having the "widget" role means that the role is not interactive
+						if (!elementToRole.has("widget") && tabIndexValue > -1) {
+							path.context.addNodeDiagnostic(
+								node,
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts:82]:
```suggestion
								tabIndexAttribute,
```

File: packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+						}
+					} else {
+						/**
+						 * Here we need to check the value of the index because
+						 * <div tabIndex="-1" ></div>
+						 * is a valid statement
+						 */
+						if (tabIndexValue > -1) {
+							path.context.addNodeDiagnostic(
+								node,
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts:94]:
```suggestion
								tabIndexAttribute,
```

File: packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+					if (role) {
+						/**
+						 * Some roles are not interactive (e.g. article),
+						 * which means that they can have a tabIndex=-1 which removes the focus behaviour
+						 *
+						 * e.g. <article tabIndex="-1" /> is a valid statement
+						 */
+						if (!isRoleInteractive(role) && tabIndexValue > -1) {
+							path.context.addNodeDiagnostic(
+								node,
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts:112]:
```suggestion
								attr,
```

File: packages/@romejs/diagnostics/descriptions/lint.ts [hunk 6]
```diff
@@ -16,6 +16,17 @@ import {buildSuggestionAdvice} from "../helpers";
 import {createDiagnosticsCategory, orJoin} from "./index";
 
 export const lint = createDiagnosticsCategory({
+	JSX_A11Y_NO_NONINTERACTIVE_ELEMENT_TO_INTERACTIVE_ROLE: (element: string) => ({
+		category: "lint/jsx-a11y/noNoninteractiveElementToInteractiveRole",
+		message: `The HTML element <emphasis>${element}</emphasis> is non-interactive and should not have an interactive role`,
```
Comment (bitpshr) [inline packages/@romejs/diagnostics/descriptions/lint.ts:21]:
```suggestion
		message: `The HTML element <emphasis>${element}</emphasis> is non-interactive and should not have an interactive role`.,
```

File: packages/@romejs/diagnostics/descriptions/lint.ts [hunk 7]
```diff
@@ -16,6 +16,17 @@ import {buildSuggestionAdvice} from "../helpers";
...
+		advice: [
+			{
+				type: "log",
+				category: "info",
+				text: `Consider replacing <emphasis>${element}</emphasis> with "div" or "span"`,
```
Comment (bitpshr) [inline packages/@romejs/diagnostics/descriptions/lint.ts:26]:
```suggestion
				text: `Replace <emphasis>${element}</emphasis> with a div or a span.`,
```

File: packages/@romejs/diagnostics/descriptions/lint.ts [hunk 8]
```diff
@@ -27,6 +38,10 @@ export const lint = createDiagnosticsCategory({
 			},
 		],
 	},
+	JSX_A11Y_NO_REDUNDANT_ROLES: (role: string, element: string) => ({
+		category: "lint/jsx-a11y/noRedundantRoles",
+		message: `Using the role attribute <emphasis>${role}</emphasis> and the HTML element <emphasis>${element}</emphasis> is redundant, using the HTML element is enough.`,
```
Comment (bitpshr) [inline packages/@romejs/diagnostics/descriptions/lint.ts:43]:
```suggestion
		message: `Using the role attribute <emphasis>${role}</emphasis> on the <emphasis>${element}</emphasis> element is redundant.`,
```

File: packages/@romejs/diagnostics/descriptions/lint.ts [hunk 9]
```diff
@@ -38,6 +53,10 @@ export const lint = createDiagnosticsCategory({
 			},
 		],
 	}),
+	JSX_A11Y_NO_NONINTERACTIVE_TABINDEX: {
+		category: "lint/jsx-a11y/noNoninteractiveTabindex",
+		message: "Do not use <emphasis>tabIndex</emphasis> on an element that is not interactive",
```
Comment (bitpshr) [inline packages/@romejs/diagnostics/descriptions/lint.ts:58]:
```suggestion
		message: "Do not use <emphasis>tabIndex</emphasis> on an element that is not interactive.",
		advice: [
			{
				type: "log",
				category: "info",
				text: "Adding non-interactive elements to the keyboard navigation flow can confuse users.",
			},
		],
```

File: packages/@romejs/js-ast-utils/getJSXElementName.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+export default function getJSXElementName(node: JSXElement): string {
+	if (node.name.type === "JSXIdentifier") {
+		return node.name.name;
+	}
+	if (node.name.type === "JSXReferenceIdentifier") {
+		return node.name.name;
+	}
+	if (node.name.type === "JSXMemberExpression") {
+		// TODO need to understand what to do here, what's the case
+		return "";
```
Comment (bitpshr) [inline packages/@romejs/js-ast-utils/getJSXElementName.ts:12]: The case here is a JSX tag like `<Foo.bar>foo</Foo.bar>`. We could return `"Foo.bar"` in this case, but I'm not sure how useful this actually us. I don't think this needs to be handed at all right now.

File: packages/@romejs/compiler/lint/rules/ariaHelpers.ts [hunk 11]
```diff
@@ -3,6 +3,76 @@ export type MapOfARIAPropertyDefinitions = Map<
...
+	| "presentation"
+	| "region"
+	| "row"
+	| "rowgroup"
+	| "rowheader"
+	| "separator"
+	| "table"
+	| "term"
+	| "toolbar";
+
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/ariaHelpers.ts:75]: The additions throughout this file look semantically correct. Do you think it would be worth splitting this file into smaller modules, and putting them in an `rules/aria/` folder? This file is getting large and I don't think we've covered all roles yet.

**Comment by ematipico** [comment]
@bitpshr I addressed all your suggestions!

I'd prefer to implement/add more test cases on a second PR if possible! This PR is already big enough and it adds lot of utilities and changes that could be used for other rules.

File: packages/@romejs/compiler/lint/rules/ariaHelpers.ts (hunk 11 above)
Comment (ematipico) [reply packages/@romejs/compiler/lint/rules/ariaHelpers.ts:75]: Yes, in my opinion it's worth splitting up this file in different different files under a folder, I already do that it will get bigger and bigger! I can do that in another PR.
